lib/clib/Simpl_Rewrite.thy

```diff
@@ -276,225 +300,219 @@ lemmas never_continues_Seqs =
   by (auto simp: never_continues_def elim: exec_elim_cases)
 
 lemma never_continues_Guard:
-  "never_continues t c \<Longrightarrow> never_continues t (Guard f b c)"
+  "never_continues_if t c \<Longrightarrow> never_continues_if t (Guard f b c)"
   by (auto simp: never_continues_def elim: exec_elim_cases)
 
 lemma never_continues_Catch:
-  "never_continues tc c \<Longrightarrow> never_continues th h \<Longrightarrow> never_continues (tc \<and> th) (Catch c h)"
+  "never_continues_if tc c \<Longrightarrow> never_continues_if th h \<Longrightarrow> never_continues_if (tc \<and> th) (Catch c h)"
   by (auto simp: never_continues_def elim: exec_elim_cases)
 
 text \<open>If all else fails...\<close>
 
 lemma never_continues_False:
-  "never_continues False c"
+  "never_continues_if False c"
   by (simp add: never_continues_def)
 
-end
-
-text \<open>One layer of context around a Simpl program.
-
-      For example, if the current focus is the first branch of a @{term Cond},
-      then the context consists of a constructor @{text CondTC} which indicates
-      this is the context, and carries everything but but the focused branch,
-      i.e. the condition and the second branch.
-
-      The @{text CondFC} and @{text HandlerC} cases each carry an extra bit of
-      information, which indicates whether the left-hand branch @{text never_continues}.
-      The assumption here is that we always traverse Simpl programs left-to-right.
-      @{text Seq2C} does not require an extra bit, because we never enter the second
-      command of a @{term Seq} if the first command @{text never_continues}.\<close>
-
-datatype ('s,'p,'f) com_ctxt
-  = Seq1C "('s,'p,'f) com" \<comment> \<open>first command of a @{term Seq}\<close>
-  | Seq2C "('s,'p,'f) com" \<comment> \<open>second command of a @{term Seq}\<close>
-  | CondTC "'s bexp" "('s,'p,'f) com" \<comment> \<open>first branch of a @{term Cond}\<close>
-  | CondFC "'s bexp" "('s,'p,'f) com" bool \<comment> \<open>second branch of a @{term Cond}\<close>
-  | WhileC "'s bexp" \<comment> \<open>body of a @{term While}\<close>
-  | WhileAnnoC "'s bexp" "'s assn" "('s \<times> 's) assn" \<comment> \<open>body of a @{term whileAnno}\<close>
-  | GuardC "'f" "'s bexp" \<comment> \<open>body of a @{term Guard}\<close>
-  | TryC "('s,'p,'f) com" \<comment> \<open>body of a @{term Catch}\<close>
-  | HandlerC "('s,'p,'f) com" bool \<comment> \<open>handler of a @{term Catch}\<close>
-
 text \<open>Rewrite Simpl programs under a predicate @{term P} which is preserved by @{term com_eq}.\<close>
 
-locale simpl_rewrite = simpl_rewrite_base \<Gamma>
-  for \<Gamma> :: "'p \<Rightarrow> ('s,'p,'f) com option" +
+definition com_eq_hom :: "(('s,'p,'f) com \<Rightarrow> bool) \<Rightarrow> bool" where
+  "com_eq_hom P \<equiv> \<forall>c c'. (c \<sim> c') \<longrightarrow> P c' \<longrightarrow> P c"
```

Comment:
what does `com_eq_hom` stand for? com program equivalent ... ?

Reply:
I was thinking it's roughly a homomorphism from `com_eq` to  predicate `P`. That is, if `c` and `c'` are `com`-equivalent, then `P c` iff `P c'`. But perhaps "homomorphism" is a bit obscure. Can you think of something clearer?

Reply:
I suppose a comment to that effect would help.

Reply:
Oh, that's obscure but ultimately survivable. I think I didn't get to "homomorphism" from "hom", that's all.